### PR TITLE
Implement object tree helper module

### PR DESCRIPTION
**Pull request by EduardKaverinskyi**

### CODE REVIEW

**Review by pawtom — CHANGES_REQUESTED** [review]

File: squape/object_tree.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    Examples:
+        children(object, {'type' : Button, 'visible' : True})
+    """
+    if selector is None:
+        selector = {}
+    children = object.children(obj)
+    return tuple(filter(lambda x: _filter_by_selector(x, selector), children))
+
+
+def find(obj, selector: dict = None, max_depth=None, _depth=0) -> tuple:
```
Comment (Adakar) [inline squape/object_tree.py:35]: 1. same comments to naming, parameter type, and docstrings as previously
2. I think you don't need the `_depth` parameter at all, instead you could decrease max_depth for every recursive call.

File: squape/object_tree.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        children.extend(find(child, max_depth=max_depth, _depth=_depth + 1))
+
+    if _depth == 0:
+        filtered_children = filter(lambda x: _filter_by_selector(x, selector), children)
+        return tuple(filtered_children)
+    else:
+        return children
+
+
+def find_parent(obj, selector: dict = None):
```
Comment (Adakar) [inline squape/object_tree.py:75]: 1. same comments to naming, parameter type, and docstrings as previously

File: squape/object_tree.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    """Find the first parent that matches the selector,
```
Comment (Adakar) [inline squape/object_tree.py:76]:
```suggestion
    """Find the first ancestor that matches the selector,
```

File: squape/object_tree.py (hunk 2 above)
Comment (Adakar) [inline squape/object_tree.py:75]: @pawtom how about
```suggestion
def find_ancestor(obj, selector: dict = None):
```

File: squape/object_tree.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    if object.parent(obj) is None:
+        return None
+    
+    if _filter_by_selector(object.parent(obj), selector):
+        return object.parent(obj)
+        
+    return find_parent(object.parent(obj), selector)
+
+
+def siblings(obj, selector: dict = None):
```
Comment (Adakar) [inline squape/object_tree.py:104]: same comments to naming, parameter type, and docstrings as previously

File: squape/object_tree.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+    if parent is None:
+        return None
+    else:
+        siblings = list(object.children(parent))
+        siblings.remove(obj)
+        return tuple(filter(lambda x: _filter_by_selector(x, selector), siblings))
+
+
+def _filter_by_selector(obj, selector: dict) -> bool:
```
Comment (Adakar) [inline squape/object_tree.py:133]: 1. same comments to naming, parameter type, and docstrings as previously
2. the name of the function suggests that some kind of filtering is done here which is not the case I believe

**Review by Adakar — CHANGES_REQUESTED** [review]
@pawtom could you please x-check my comments before any further work is done here?

**Comment by pawtom** [comment]
I agree with @Adakar suggestions :)

**Review by Adakar — CHANGES_REQUESTED** [review]

File: squape/object_tree.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+    for child in object.children(object_reference):
+        if _is_matching(child, selector):
+            children += (child,)
+        children += find(child, selector, max_depth - 1)
+
+    return children
+
+
+def find_ancestor(object_name: any, selector: dict = None):
```
Comment (Adakar) [inline squape/object_tree.py:88]: @pawtom shall we make the selector a  required parameter? Without selector shit function behaves like `parent()`

File: squape/object_tree.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    """
+    if selector is None:
+        selector = {}
+    object_reference = _get_object_reference(object_name)
+    children = object.children(object_reference)
+    return tuple(filter(lambda x: _is_matching(x, selector), children))
+
+
+def find(object_name: any, selector: dict = None, max_depth: int = None) -> tuple:
+    """Recursively filter all the children of the specified object
```
Comment (Adakar) [inline squape/object_tree.py:44]:
```suggestion
    """Finds all the children of the given object.
```

**Review by Adakar — COMMENTED** [review]
@pawtom Could you please get familiar with the comments I made and share your thoughts?

Besides that, the functions that user can pass as a part of selector don't have to be lambdas. They can use function references instead:
```python
def foo(x):
    return True if x>56 else False

object_tree.children(names.address_Book_Add_Dialog, {"x": foo})
```
It's a case similar to the `key` optional parameter of the [sort()](https://docs.python.org/3/library/stdtypes.html#list.sort) function. It's documentation may give you same ideas for documenting this feature in a nicer way. I don't think we should use **lambda**.

File: squape/object_tree.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+try:
+    import squish
+except ImportError:
+    import squishtest as squish
+
+import object
+
+
+def children(object_name: any, selector: dict = None) -> tuple:
```
[the remark this answers is not in the document]
Comment (pawtom) [reply squape/object_tree.py:18]: Not sure if I agree, I want the function to be versatile.

File: squape/object_tree.py (hunk 7 above)
Comment (pawtom) [reply squape/object_tree.py:44]: "Finds all descendants of the given object. "

File: squape/object_tree.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Adakar) [reply squape/object_tree.py:18]: @pawtom It's the same case like find_ancestor(). 
using squpe.find_ancestor() with no selector doesn't make sense, user should use Squish's object.parent() instead.
using squape.children() with no selector doesn't make sense, user should use Squish's object.children() instead.

using squape in such cases is inefficient and users should use Squish API anyway

**Comment by EduardKaverinskyi** [comment]
Another suggestion is to add handling of the case when a provided function does not accept any parameters or expects more than 1 parameter. We should handle this and give the user an explicit message of what is wrong.

File: squape/object_tree.py (hunk 6 above)
Comment (Adakar) [reply squape/object_tree.py:88]: @EduardKaverinskyi please make the selector parameter required

File: squape/object_tree.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Adakar) [reply squape/object_tree.py:18]: @EduardKaverinskyi please make the selector parameter required

File: squape/object_tree.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    """Filter the direct children of the specified object by the specified selector
+
+    Args:
+        object_name (any): symbolic name, real name, or object reference to the parent
+        selector (dict, optional): The selector is a dictionary of properties,
+        that must match for objects to be included into the result. 
+        The values of the selector can be lambda functions as well.
```
[the remark this answers is not in the document]
Comment (Adakar) [reply squape/object_tree.py:25]: lambda is not a function reference, please use just "function"

**Comment by Adakar** [comment]
> Another suggestion is to add handling of the case when a provided function does not accept any parameters or expects more than 1 parameter. We should handle this and give the user an explicit message of what is wrong.

@EduardKaverinskyi please elaborate and give some example? What function do you have in mind?

**Comment by EduardKaverinskyi** [comment]
> > Another suggestion is to add handling of the case when a provided function does not accept any parameters or expects more than 1 parameter. We should handle this and give the user an explicit message of what is wrong.
> 
> @EduardKaverinskyi please elaborate and give some example? What function do you have in mind?

Case 1:
lambda: 1/2
This function accepts 0 parameters, but we always pass 1. It will throw an unhandled error

Case 2:
lambda x,y: x*y
This function accepts 2 parameters, but we always pass 1. It will throw and unhabndled error.

Please not that those examples are very simplified for understading. Actual functions may be much more complicated, but the idea remains the same.

The same behavior will happen if we pass function references instead of lambda functions. For both cases we should handle it.

**Comment by Adakar** [comment]
This sounds like a good idea but may be a bit overkill. I'd leave is as it is - that seems to be approach used by standard lib functions like sort().

File: squape/object_tree.py [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+
+    Args:
+        object_name (any): symbolic name, real name, or object reference to the parent.
+
+        selector (dict): The selector is a dictionary of key-value pairs,
+        where a key is a property of an object, and value is expected value.
+        The object will pass verification
+        if the object's property value matches the selector value.
+        Selectors also may include functions.
+        The passed functions must accept exactly 1 argument.
```
Comment (Adakar) [inline squape/object_tree.py:30]:
```suggestion
        selector (dict): The selector is a dictionary of key-value pairs,
            where a key is a property of an object, and value is expected value.
            An object will pass verification
            if the object's property value matches the selector value.
            Selectors may include functions.
            The passed functions must accept exactly one argument.
```

**Review by Adakar — CHANGES_REQUESTED** [review]
@EduardKaverinskyi after reading doc strings again I improved wording a bit.  Could you please use it in other functions as well? Sorry for that.
You forgot to fix indentation. Please do that for other functions as well.

File: squape/object_tree.py (hunk 10 above)
Comment (EduardKaverinskyi) [reply squape/object_tree.py]: Ah, I misunderstood a bit. Now it is clear. Please see my last commit.

**Review by Adakar — APPROVED** [review]
Looks good :)

**Pull request merged**